tests/unit/orchestrate.py

```diff
@@ -0,0 +1,356 @@
+    number of turns it takes to run each test to completion. In this example, it will
+    take two turns for each test: one turn to start the thread and execute up until the
+    syncpoint, and then another turn to execute from the syncpoint to the end of the
+    test. The number of turns will always be one greater than the number of syncpoints
+    encountered when executing the test.
+
+    Once the counts have been taken, `orchestrate` will construct a test sequence that
+    represents the all the turns taken by the passed in tests, with each value in the
+    sequence representing the the index of the test whose turn it is in the sequence. In
+    this example, then, it would produce::
+
+        [0, 0, 1, 1]
+
+    This represents the first test taking both of its turns, followed by the second test
+    taking both of its turns. At this point this scenario has already been tested,
+    because this is what was run to produce the counts and the initial test sequence.
+    Now `orchestrate` will run all of the remaining scenarios by finding all the
+    permutations of the test sequence and executing those, in turn::
+
+        [0, 1, 0, 1]
+        [0, 1, 1, 0]
+        [1, 0, 0, 1]
+        [1, 0, 1, 0]
+        [1, 1, 0, 0]
+
+    You'll notice in our example that since both test functions are actually the same
+    function, that although it tested 6 scenarios there are effectively only really 3
+    unique scenarios. For the time being, though, `orchestrate` doesn't attempt to
+    detect this condition or optimize for it.
+
+    There are some performance considerations that should be taken into account when
+    writing tests. The number of unique test sequences grows quite quickly with the
+    number of turns taken by the functions under test. Our simple example with two
+    threads each taking two turns, only yielded 6 scenarios, but two threads each taking
+    6 turns, for example, yields 924 scenarios. Add another six step thread and now you
+    have over 17 thousand scenarios. In general, use the least number of steps/threads
+    you can get away with and still expose the behavior you want to correct.
+
+    For the same reason as above, if you have many concurrent tests, when writing a new
+    test, make sure you're not accidentally patching syncpoints intended for other
+    tests, as this will add steps to your tests. While it's not problematic from a
+    testing standpoint to have extra steps in your tests, it can use computing resources
+    unnecessarily. Using different no-op functions with different names for different
+    tests can help with this.
+
+    As soon as any error or failure is detected, no more scenarios are run
+    and that error is propagated to the main thread.
+
+    Args:
+        tests (Tuple[Callable]): Test functions to be run. These functions will not be
+            called with any arguments, so they must not have any required arguments.
```

Comment:
Would it be useful to add `**kwargs` here, and plumb it down through the `_TestThread`?  I can imagine wanting to `@pytest.mark.parametrize` with different keyword values to exercise different cases.

Or maybe that is just over-engineering here, especially given my "brain-bending" comment above.

Reply:
I think between closures and `functools.partial`, people have plenty of options for producing arg-free callables.